backend/src/apiserver/server/util.go

```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
+func ValidatePipelineSpecAndResourceReferences(resourceManager *resource.ResourceManager, spec *api.PipelineSpec, resourceReferences []*api.ResourceReference) error {
+	// centralize all the nil checks here
+	pipelineId := spec.GetPipelineId()
+	workflowManifest := spec.GetWorkflowManifest()
+	pipelineVersionId := getPipelineVersionIdFromResourceReferences(resourceManager, resourceReferences)
+
+	if err := validateParameters(spec.GetParameters()); err != nil {
+		return err
+	}
+	if workflowManifest != "" {
+		if pipelineId != "" || pipelineVersionId != "" {
+			return util.NewInvalidInputError("Please don't specify a pipeline version or pipeline ID when you specify a workflow manifest.")
+		}
+		if err := validateWorkflowManifest(spec.GetWorkflowManifest()); err != nil {
+			return err
+		}
+	} else {
+		if pipelineId == "" && pipelineVersionId == "" {
+			return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
+		}
+		if err := validatePipelineId(resourceManager, pipelineId); err != nil {
+			return err
+		}
+		if pipelineVersionId != "" {
+			// verify pipelineVersionId exists
+			pipelineVersion, err := resourceManager.GetPipelineVersion(pipelineVersionId)
+			if err != nil {
+				return util.Wrap(err, "Get pipelineVersionId failed.")
+			}
+			// verify pipelineId should be parent of pipelineVersionId
+			if pipelineId != "" && pipelineVersion.PipelineId != pipelineId {
+				return util.NewInvalidInputError("pipeline ID should be parent of pipeline version.")
+			}
+		}
 	}
-	if spec.GetPipelineId() != "" && spec.GetWorkflowManifest() != "" {
-		return util.NewInvalidInputError("Please either specify a pipeline ID or a workflow manifest, not both.")
+	return validateParameters(spec.GetParameters())
+}
+func validateParameters(parameters []*api.Parameter) error {
+	if parameters != nil {
+		paramsBytes, err := json.Marshal(parameters)
+		if err != nil {
+			return util.NewInternalServerError(err,
+				"Failed to Marshall the pipeline parameters into bytes. Parameters: %s",
+				printParameters(parameters))
+		}
+		if len(paramsBytes) > util.MaxParameterBytes {
+			return util.NewInvalidInputError("The input parameter length exceed maximum size of %v.", util.MaxParameterBytes)
+		}
 	}
-	if spec.GetPipelineId() != "" {
+	return nil
+}
+
+func validatePipelineId(resourceManager *resource.ResourceManager, pipelineId string) error {
+	if pipelineId != "" {
 		// Verify pipeline exist
-		if _, err := resourceManager.GetPipeline(spec.GetPipelineId()); err != nil {
-			return util.Wrap(err, "Get pipeline failed.")
+		if _, err := resourceManager.GetPipeline(pipelineId); err != nil {
+			return util.Wrap(err, "Get pipelineId failed.")
 		}
 	}
-	if spec.GetWorkflowManifest() != "" {
+	return nil
+}
+
+func validateWorkflowManifest(workflowManifest string) error {
+	if workflowManifest != "" {
```

Comment:
nit: although the `workflowManifest == ""` branch may never be reached here

I think it makes more sense to return an error, because when we validate workflow manifest, empty manifest should be an error.

Reply:
I don't think empty manifest should be an error. For example,  workflow manifest can be empty when we have pipeline id. Same logic used when we validatePipelineId.

Reply:
Whether empty manifest is valid or not depends on other variables. We won't know it when we validate workflow manifest itself and I don't think we should include the logic when we validate workflow manifest itself.

Reply:
I'm thinking about a different definition of what this function does -- "when we have already decided to use workflow manifest, is it valid?". My rationale is that, we've already handled the complexity of when to choose which pipeline source outside. So when we actually call `validateWorkflowManifest`, we know we will be using workflow manifest as pipeline source and empty is not a valid workflow manifest.

The difference is very nuanced and the code path won't be reached, so above is just explaining my thoughts. You can make the decision.